service-base/src/main/java/org/eclipse/hono/service/http/AbstractHttpEndpoint.java

```diff
@@ -201,6 +205,58 @@ protected final String getDeviceIdParam(final RoutingContext ctx) {
         return ctx.request().getParam(PARAM_DEVICE_ID);
     }
 
+    /**
+     * Gets the value of a request parameter.
+     *
+     * @param ctx The routing context to get the parameter from.
+     * @param paramName The name of the parameter.
+     * @param validator A predicate to use for validating the parameter value or
+     *                  {@code null} if the value should not be validated.
+     * @param optional {@code true} if parameter is not mandatory.
+     * @return A future indicating the outcome of the operation.
+     *         If the request does not contain a parameter with the given name, the future will be
+     *         <ul>
+     *         <li>completed with an empty optional if the <em>optional</em> flag is {@code true}, or</li>
+     *         <li>failed with a {@link ClientErrorException} with status 400 if the flag is {@code false}.</li>
+     *         </ul>
+     *         If the request does contain a parameter with the given name, the future will be
+     *         <ul>
+     *         <li>failed with a {@link ClientErrorException} with status 400 if a predicate has been
+     *         given and the predicate evaluates to {@code false}, or</li>
+     *         <li>otherwise be completed with the parameter value.</li>
+     *         </ul>
+     * @throws NullPointerException If ctx or paramName are {@code null}.
+     */
+    protected final Future<Optional<String>> getRequestParameter(
+            final RoutingContext ctx,
+            final String paramName,
+            final Pattern validator,
```

Comment:
What if we make validator a function? Then you can pass any logic to verify your parameter value, like pattern matching, optional, etc. It would make 'optional' field obsolete as well.

Reply:
That might work if the return type is a `Future<String>` so that we can handle the cause for failures properly. Let me think about it ...